libs/agent-sm/agent-policy/src/main/java/org/opensearch/secure_sm/policy/PolicyFile.java

```diff
@@ -0,0 +1,346 @@
+        init(url, newInfo);
+    }
+
+    private void init(URL policy, PolicyInfo newInfo) throws PolicyInitializationException {
+        try (InputStreamReader reader = new InputStreamReader(getInputStream(policy), StandardCharsets.UTF_8)) {
+            PolicyParser policyParser = new PolicyParser();
+            policyParser.read(reader);
+
+            for (GrantEntry grantEntry : policyParser.grantElements()) {
+                addGrantEntry(grantEntry, newInfo);
+            }
+
+        } catch (Exception e) {
+            throw new PolicyInitializationException("Failed to load policy from: " + policy, e);
+        }
+    }
+
+    public static InputStream getInputStream(URL url) throws IOException {
+        if ("file".equals(url.getProtocol())) {
+            String path = url.getFile().replace('/', File.separatorChar);
+            path = URLDecoder.decode(path, StandardCharsets.UTF_8);
+            return new FileInputStream(path);
+        } else {
+            return url.openStream();
+        }
+    }
+
+    private CodeSource getCodeSource(GrantEntry grantEntry, PolicyInfo newInfo) throws PolicyInitializationException {
+        try {
+            Certificate[] certs = null;
+            URL location = (grantEntry.codeBase != null) ? newURL(grantEntry.codeBase) : null;
+            return canonicalizeCodebase(new CodeSource(location, certs));
+        } catch (Exception e) {
+            throw new PolicyInitializationException("Failed to get CodeSource", e);
+        }
+    }
+
+    private void addGrantEntry(GrantEntry grantEntry, PolicyInfo newInfo) throws PolicyInitializationException {
+        CodeSource codesource = getCodeSource(grantEntry, newInfo);
+        if (codesource == null) {
+            throw new PolicyInitializationException("Null CodeSource for: " + grantEntry.codeBase);
+        }
+
+        PolicyEntry entry = new PolicyEntry(codesource);
+        List<PermissionEntry> permissionList = grantEntry.permissionElements();
+        for (PermissionEntry pe : permissionList) {
+            expandPermissionName(pe);
+            try {
+                Optional<Permission> perm = getInstance(pe.permission, pe.name, pe.action);
+                if (perm.isPresent()) {
+                    entry.add(perm.get());
+                }
+            } catch (ClassNotFoundException e) {
+                // these were mostly custom permission classes added for security
+                // manager. Since security manager is deprecated, we can skip these
+                // permissions classes.
+                if (PERM_CLASSES_TO_SKIP.contains(pe.permission)) {
+                    continue; // skip this permission
+                }
+                throw new PolicyInitializationException("Permission class not found: " + pe.permission, e);
+            }
+        }
+        newInfo.policyEntries.add(entry);
+    }
+
+    private void expandPermissionName(PermissionEntry pe) {
```

Comment:
I guess this is necessary as well? I'm still confused at the usage of this method. What would be an example polify file where the permission name is expanded?

Reply:
that's be needed for something like -- ```permission java.io.FilePermission "${{prop:user.kumargu_directory}}", "read";```

I think you had evaluated that we have no occurrences of such policies, but I was just not comfortable to remove it (due to flaky tests)

Reply:
oh, I had this mixed up. I see that PermissionEntry is 3 attributes:

1. permission -> the name of the java class
2. name
3. action

I think I was confusing number 1 and 2 together. Disregard my comment. Marking as resolved.